File: go/leetcode/5-longest-palindromic-substring.go
package main

func longestPalindrome(s string) string {
	n := len(s)

	if n < 2 {
		return s
	}

	start, end := 0, 0

	expand := func(l, r int) (int, int) {
		for l >= 0 && r < n && s[l] == s[r] {
			l--
			r++
		}

		// Кэшируем последнее подходящее значение (шаг границ назад)
		return l + 1, r - 1
	}

	for i := range n {
		l1, r1 := expand(i, i)
		l2, r2 := expand(i, i+1)

		if r1-l1 > end-start {
			start, end = l1, r1
		}
		if r2-l2 > end-start {
			start, end = l2, r2
		}
	}

	return s[start : end+1]
}

File: ts/leetcode/5-longest-palindromic-substring.ts
function longestPalindrome(s: string): string {
    if (s.length < 2) return s;

    let start = 0, maxLen = 1;

    function expand(l: number, r: number) {
        while (l >= 0 && r < s.length && s[l] === s[r]) {
            const len = r - l + 1;

            if (len > maxLen) {
                start = l; // start становится левой границей
                maxLen = len;
            }

            l--;
            r++;
        }
    }

    for (let i = 0; i < s.length; i++) {
        expand(i, i + 1);     // Проверка с четным центром
        expand(i - 1, i + 1); // Проверка с нечетным центром
    }

    return s.slice(start, start + maxLen);
}